src/history/hs_rec.c

```diff
@@ -152,8 +166,12 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
         if (ret == 0)
             WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
         else {
+            F_SET(cursor, WT_CURSTD_HS_READ_ALL);
+
             cursor->set_key(cursor, 3, btree->id, key, tw->start_ts + 1);
             WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, cursor), true);
+
+            F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
```

Comment:
I think we should restore the previous flag setting instead of just clearing the flag.

Reply:
Fixed.